### PR TITLE
Introduce black formatter

### PR DESCRIPTION
**Pull request by awav**
- [x] Format GPflow
- [x] Add checker to the CI

Optional:
- [ ] Add auto-reformat as the [git hook](https://www.mattlayman.com/blog/2018/python-code-black/)?

### CODE REVIEW

File: gpflow/conditionals/uncertain_conditionals.py [hunk 1]
```diff
@@ -59,60 +63,61 @@ def uncertain_conditional(Xnew_mu: tf.Tensor,
     if not white:
         q_mu = tf.linalg.triangular_solve(Luu, q_mu, lower=True)
         Luu_tiled = tf.tile(
-            Luu[None, :, :],
-            [num_func, 1, 1])  # remove line once issue 216 is fixed
+            Luu[None, :, :], [num_func, 1, 1]
+        )  # remove line once issue 216 is fixed
```
Comment (joelberkeley-pio) [inline gpflow/conditionals/uncertain_conditionals.py]: comment now points to wrong line

File: gpflow/conditionals/uncertain_conditionals.py [hunk 2]
```diff
@@ -59,60 +63,61 @@ def uncertain_conditional(Xnew_mu: tf.Tensor,
...
 
     Li_eKuf = tf.linalg.triangular_solve(Luu, eKuf, lower=True)  # [M, N]
     fmean = tf.linalg.matmul(Li_eKuf, q_mu, transpose_a=True)
 
     eKff = expectation(pXnew, kernel)  # N (psi0)
-    eKuffu = expectation(pXnew, (kernel, inducing_variable),
-                         (kernel, inducing_variable))  # [N, M, M] (psi2)
+    eKuffu = expectation(
+        pXnew, (kernel, inducing_variable), (kernel, inducing_variable)
+    )  # [N, M, M] (psi2)
```
Comment (joelberkeley-pio) [inline gpflow/conditionals/uncertain_conditionals.py]: comment now points to wrong line

**Comment by joelberkeley-pio** [comment]
contrary to my comments about black before, I think that, even though there are a few places I don't agree with the format style, that this is a good thing. I'm completely pro using it.

One question: can you tell it to ignore certain sections of code if there's context that black can't pick up on

**Comment by awav** [comment]
> One question: can you tell it to ignore certain sections of code if there's context that black can't pick up on?

As far as I know no

**Comment by st--** [comment]
> One question: can you tell it to ignore certain sections of code if there's context that black can't pick up on?

Yes, you can. From the [Black docs](https://black.readthedocs.io/en/stable/the_black_code_style.html#the-black-code-style):
> It doesn’t reformat blocks that start with # fmt: off and end with # fmt: on. # fmt: on/off have to be on the same level of indentation. It also recognizes YAPF’s block comments to the same effect, as a courtesy for straddling code.

File: gpflow/conditionals/uncertain_conditionals.py (hunk 1 above)
Comment (st--) [reply gpflow/conditionals/uncertain_conditionals.py]: you just have to read "line" more generously...

**Comment by awav** [comment]
Thanks, @st--! @joelberkeley-pio where do you think we need to turn off reformatting?

**Comment by joelberkeley-pio** [comment]
> Thanks, @st--! @joelberkeley-pio where do you think we need to turn off reformatting?

Nowhere now, but I didn't want us to end up constrained by in the rare case we do need unrestrained formatting

File: contributing.md [hunk 3]
```diff
@@ -63,6 +63,14 @@ class Foo:
         return session.run(incr_tensor)
 ```
 
+### Formatting
+
+GPflow uses [black](https://github.com/psf/black) for formatting in the following way:
+
+```bash
+black -t py36 --line-length=100 gpflow/
```
Comment (joelberkeley-pio) [inline contributing.md]: worth formatting the tests too?

**Comment by joelberkeley-pio** [comment]
as for the PR, can you explain the changes? there are 62 files of changes. which ones are infrastructure changes and which are formatting changes made by black?

**Comment by awav** [comment]
@joelberkeley-pio, all *.py files changed by `black`, the rest is CI and doc changes.

File: contributing.md (hunk 3 above)
Comment (awav) [inline contributing.md]:
```suggestion
black -t py36 -l 100 gpflow tests
```

File: .circleci/config.yml [hunk 4]
```diff
@@ -33,6 +33,21 @@ jobs:
         environment:
           PYTEST_FILTER: notebooks
 
+  format-checker:
+    docker:
+      - image: awav/tensorflow:2.0-preview
```
Comment (vdutor) [inline .circleci/config.yml]: when are we upgrading this to awav/tensorflow:2.1 ?

Comment (awav) [reply .circleci/config.yml]: The 2.0-preview doesn't mean it uses 2.0 version. Docker image doesn't have pre-installed tensorflow, we install things via pip. I will address the image name in different PR.

**Pull request merged**